src/Npgsql/Internal/NpgsqlConnector.cs

```diff
@@ -686,25 +697,31 @@ async ValueTask<string> GetUsernameAsync(bool async, CancellationToken cancellat
     {
         var username = Settings.Username;
         if (username?.Length > 0)
-            return username;
+            return SetInferredUserName(username);
```

Comment:
Nit: factoring-wise this could be a simple local async function which returns the username, and the outer one (GetUserNameAsync) could be a non-async one which manages the storing (this would obviate the SetInferredUserName function). We can also choose to cache the value rather than e.g. doing a Kerberos lookup each time (if we assume it won't change during the process lifetime, which it most probably won't).

Reply:
> Nit: factoring-wise this could be a simple local async function which returns the username, and the outer one (GetUserNameAsync) could be a non-async one which manages the storing (this would obviate the SetInferredUserName function).

How can we get hold of the username in the outer function if it isn't async? We would get a `ValueTask<string>` from the inner local function wich we cannot await (being non-async).

> We can also choose to cache the value rather than e.g. doing a Kerberos lookup each time (if we assume it won't change during the process lifetime, which it most probably won't).

The `KerberosUsernameProvider` actually caches already and I don't think that we should easily cache the values we infer from other sources here. But see my general thoughts about caching inferred values at https://github.com/npgsql/npgsql/pull/4765#discussion_r1024843788

Reply:
> How can we get hold of the username in the outer function if it isn't async? We would get a ValueTask<string> from the inner local function wich we cannot await (being non-async).

I just meant it could theoretically be optimized into only having a state machine if the value hasn't been cached yet (if the cached value is populated, return it without any I/O or async, otherwise call the local method and return the ValueTask it returns).

But it's probably not worth micro-optimizing (unless we think that e.g. the Kerberos user lookup could be problematic). As a nit I'd still maybe refactor just to avoid having SetInferredUserName but it's not important.

Reply:
> I just meant it could theoretically be optimized into only having a state machine if the value hasn't been cached yet (if the cached value is populated, return it without any I/O or async, otherwise call the local method and return the ValueTask it returns).

Ok, I've done this at the level of `KerberosUsernameProvider.GetUsernameAsync()` non, where we actually cache the result and where the optimization seemed obvious to me. But I don't get how you deem it possible in `NpgsqlConnector.GetUsernameAsync()` without changing our current behavior.

> As a nit I'd still maybe refactor just to avoid having SetInferredUserName but it's not important.
You mean inlining it manually?

I've done that now but I really don't get the point since it will get inlined in any case and it's just a wee bit DRYer with `SetInferredUserName()`.
If you have something else in mind I'd suggest to just commit this as it is now and you do whatever refacturing you mean afterwards. I don't have any strong feelings about this, I just don't seem to get what you mean.

Reply:
Ok, I've done one more optimization for the (most common) best case scenario but now I'm definitely out of ideas 😄

Reply:
LGTM